### PR TITLE
Nested opaque identifier

### PR DESCRIPTION
**Pull request by SoAsEr**
## Summary
This is an attempt to fix https://github.com/facebook/react/issues/20822
This allows for recursive key creation from an opaque identifier returned from useOpaqueIdentifier using a new member function called `makeNewFromKey(key: string)`. The recursive nature is useful in order to be able to identify a parent and a child. 
## Test Plan
Added a bunch of integration tests, but there are still many many edge cases that are difficult to test for.
Examples:
in the client key creation, adding "a" and then "b" should not equal "ab". Adding a simple delimiter would appear to fix this, but doesn't as then inserting the delimiter causes the same problem. Testing for this requires knowledge of the code where it places delimiters (before, after, both). In code, the above bug is avoided by appending the depth. There is a delimiter as to avoid very deep depths which require multiple digits running into the same problem as a naked delimiter. so the first is "a:0b:1" and the second is "ab:0".

in the server key creation, the multiple digit problem strikes again, and so a delimiter is added between identifiers. 

Testing every corner case like this would require huge numbers of tests, and I'm not entirely sure how to go about it. Advice would be appreciated. Removing the recursiveness entirely would fix this problem but I think it is worth putting some effort into keeping it.

### CODE REVIEW

File: packages/react-dom/src/client/ReactDOMHostConfig.js [hunk 1]
```diff
@@ -1013,14 +1012,84 @@ export function getInstanceFromNode(node: HTMLElement): null | Object {
...
+    },
+  };
+}
+
 let clientId: number = 0;
 export function makeClientId(): OpaqueIDType {
-  return 'r:' + (clientId++).toString(36);
+  const id = 'r:' + (clientId++).toString(36);
+  return {
+    makeNewFromKey(key) {
```
Comment (rheh) [inline packages/react-dom/src/client/ReactDOMHostConfig.js:1039]: Should return types be added on these?

File: packages/react-dom/src/server/ReactPartialRendererHooks.js [hunk 2]
```diff
@@ -481,12 +484,91 @@ function useTransition(): [(callback: () => void) => void, boolean] {
...
+    (currentPartialRenderer.uniqueID++).toString(36);
+  const keys: {
+    unsorted: Array<string>,
+    map: Map<string, number>,
+  } = {
+    unsorted: [],
+    map: new Map(),
+  };
+  return {
+    makeNewFromKey(key: string) {
```
Comment (rheh) [inline packages/react-dom/src/server/ReactPartialRendererHooks.js:543]: Is this section (543<563) the same as 499->527?  If so, move to a shared function?

File: packages/react-reconciler/src/ReactFiberHooks.new.js [hunk 3]
```diff
@@ -1784,20 +1786,35 @@ function warnOnOpaqueIdentifierAccessInDEV(fiber) {
...
     }
   }
 }
 
+function warnOnDuplicateOpaqueIdentifierKeyInDEV(fiber) {
+  if (__DEV__) {
+    const name = getComponentName(fiber.type) || 'Unknown';
+    if (getIsRendering() && !didWarnAboutDuplicateOpaqueIdentifierKeys[name]) {
+      console.error(
+        'Warning: Detected that two nested opaque identifiers with the same parent were created with identical keys. Keys must be unique.',
```
Comment (rheh) [inline packages/react-reconciler/src/ReactFiberHooks.new.js:1801]: Should we add name in this debug/error?

File: packages/react-reconciler/src/ReactFiberHooks.old.js [hunk 4]
```diff
@@ -1784,20 +1786,35 @@ function warnOnOpaqueIdentifierAccessInDEV(fiber) {
...
-          'passed through to attributes only. Do not read the ' +
-          'value directly.',
+        'Warning: The object passed back from useOpaqueIdentifier and those made from it are meant to be passed through to attributes only. Do not read the value directly.',
       );
       didWarnAboutUseOpaqueIdentifier[name] = true;
     }
   }
 }
 
+function warnOnDuplicateOpaqueIdentifierKeyInDEV(fiber) {
```
Comment (rheh) [inline packages/react-reconciler/src/ReactFiberHooks.old.js:1796]: Is this duplicated above?  If so, can we moved to a shared place?

File: packages/react-dom/src/client/ReactDOMHostConfig.js (hunk 1 above)
Comment (SoAsEr) [reply packages/react-dom/src/client/ReactDOMHostConfig.js:1039]: OpaqueIdType already has the return types of these functions

File: packages/react-reconciler/src/ReactFiberHooks.old.js (hunk 4 above)
Comment (SoAsEr) [reply packages/react-reconciler/src/ReactFiberHooks.old.js:1796]: .old and .new seem to share everything except imports. I didn't want to make design decisions so I just did similarly to what was around me.

File: packages/react-reconciler/src/ReactFiberHooks.new.js (hunk 3 above)
Comment (SoAsEr) [reply packages/react-reconciler/src/ReactFiberHooks.new.js:1801]: Name is just the fiber name. I just copied the similar error above it.

File: scripts/error-codes/codes.json [hunk 5]
```diff
@@ -374,5 +374,7 @@
   "383": "This query has received fewer parameters than the last time the same query was used. Always pass the exact number of parameters that the query needs.",
   "384": "Refreshing the cache is not supported in Server Components.",
   "385": "A mutable source was mutated while the %s component was rendering. This is not supported. Move any mutations into event handlers or effects.",
-  "386": "The current renderer does not support microtasks. This error is likely caused by a bug in React. Please file an issue."
+  "386": "The current renderer does not support microtasks. This error is likely caused by a bug in React. Please file an issue.",
+  "387": "Warning: The object passed back from useOpaqueIdentifier and those made from it are meant to be passed through to attributes only. Do not read the value directly.",
```
Comment (SoAsEr) [inline scripts/error-codes/codes.json:378]: Why is the `Warning:` present here? I just used the automatic lint command, I'm really not sure what I'm doing here

**Comment by eps1lon** [comment]
Bump

**Comment by SoAsEr** [comment]
New approach is totally different and doesn't need this complicated scheme: reactwg/react-18#111
